flink-runtime/src/main/java/org/apache/flink/runtime/util/AtomicDisposableReferenceCounter.java

```diff
@@ -94,9 +100,9 @@ public boolean isDisposed() {
         }
     }
 
-    public boolean disposeIfNotUsed() {
+    public boolean disposeIfNotUsed(boolean canDispose) {
         synchronized (lock) {
-            if (referenceCount <= disposeOnReferenceCount) {
+            if (referenceCount <= disposeOnReferenceCount && canDispose) {
                 isDisposed = true;
             }
```

Comment:
I don't know what this class is introduced for in the very beginning, but now it does not make sense to me any more.
- The constructor with custom `disposeOnReferenceCount` is only use in a test case.
- Without custom `disposeOnReferenceCount`, this class basically an `AtomicInteger`, where `isDisposed` can be replaced by `== 0`.
- This class introduces a constraint that the reference count can only reach the "disposed state" once. This IMO is a design flaw. It should be the caller's decision what does the counter reaching zero mean and what action to take at the time.

In summary, I think the code will be significantly simplified if we get rid of this class and replace it with a simple `AtomicInteger`, or a light-weighted wrap with shortcuts like `incAndIsZero`, `decAndIsZero`, `isZero`.